algorithms/kernel/dtrees/forest/regression/oneapi/cl_kernels/df_batch_regression_kernels.cl

```diff
@@ -0,0 +1,994 @@
+            }
+        }
+
+        barrier(CLK_LOCAL_MEM_FENCE);
+        if (1 < n_sub_groups && 0 == sub_group_id)
+        {
+            // first sub group for current node reduces over local buffer if required
+            algorithmFPType curImpDec = (sub_group_local_id < n_sub_groups) ? bufI[bufIdx + sub_group_local_id] : (algorithmFPType)0;
+
+            int curFeatureId    = sub_group_local_id < n_sub_groups ? bufS[(bufIdx + sub_group_local_id) * nNodeProp + 0] : valNotFound;
+            int curFeatureValue = sub_group_local_id < n_sub_groups ? bufS[(bufIdx + sub_group_local_id) * nNodeProp + 1] : valNotFound;
+            int LN              = sub_group_local_id < n_sub_groups ? bufS[(bufIdx + sub_group_local_id) * nNodeProp + 2] : 0;
+
+            algorithmFPType bestImpDec = sub_group_reduce_max(curImpDec);
+
+            int bestFeatureId    = sub_group_reduce_min(bestImpDec == curImpDec ? curFeatureId : valNotFound);
+            int bestFeatureValue = sub_group_reduce_min((bestFeatureId == curFeatureId && bestImpDec == curImpDec) ? curFeatureValue : valNotFound);
+
+            if (curFeatureId == bestFeatureId && curFeatureValue == bestFeatureValue)
+            {
+                __global algorithmFPType * splitNodeInfo = splitInfo + nodeId * nImpProp;
+                nodeList[nodeId * nNodeProp + 2]         = curFeatureId == valNotFound ? -1 : curFeatureId;
+                nodeList[nodeId * nNodeProp + 3]         = curFeatureValue == valNotFound ? -1 : curFeatureValue;
+                nodeList[nodeId * nNodeProp + 4]         = (int)LN;
+
+                splitNodeInfo[0] = ((algorithmFPType)0 != mrgN) ? mrgSum2Cent / mrgN : 0;
+                splitNodeInfo[1] = mrgMean;
+
+                if (updateImpDecreaseRequired) nodeImpDecreaseList[nodeId] = ((algorithmFPType)0 != mrgN) ? curImpDec / mrgN : (algorithmFPType)0;
+            }
+        }
+    }
+
+    __kernel void computeBestSplitByHistogram(const __global algorithmFPType * histograms, const __global int * selectedFeatures,
+                                              int nSelectedFeatures, const __global int * binOffsets, __global int * nodeList,
+                                              const __global int * nodeIndices, int nodeIndicesOffset, __global algorithmFPType * splitInfo,
+                                              __global algorithmFPType * nodeImpDecreaseList, int updateImpDecreaseRequired, int nMaxBinsAmongFtrs,
+                                              int minObservationsInLeafNode, algorithmFPType impurityThreshold) {
+        // this kernel has almost the same code as computeBestSplit
+        // the difference is that here for each potential split point we pass through bins hist instead of rows
+        // nodeList will be updated with split attributes in this kernel
+        // spliInfo will contain node impurity and mean
+        const int nProp              = 3; // num of characteristics in histogram
```

Comment:
I suggest to pass such kind of constants as a function parameters or at least defines to have one place which controls it (outside of the kernel).
Now you have to manually change it in few places if case of some code modifications.

Reply:
agree

Reply:
replaced them with corresponding defines